tests/kernel/wave/wave_e2e_test.py

```diff
@@ -2096,3 +2096,64 @@ def test(
 
     test(a, result_self_index)
     assert_close(ref, result_self_index[0, 0, :])
+
+
+@require_e2e
+def test_debug_log_write():
+    M = tkl.sym.M
+    N = tkl.sym.N
+    ADDRESS_SPACE = tkl.sym.ADDRESS_SPACE
+
+    shape = (256, 128)
+
+    wave_size = 64
+    BLOCK_M = 1
+    BLOCK_N = sympy.Max(sympy.Min(shape[1], 256), wave_size)
+
+    constraints: list[tkw.Constraint] = [
+        tkw.HardwareConstraint(
+            threads_per_wave=wave_size,
+            vector_shapes={M: BLOCK_M, N: BLOCK_N},
+        )
+    ]
+
+    constraints += [tkw.WorkgroupConstraint(M, BLOCK_M, 1)]
+    constraints += [tkw.WorkgroupConstraint(N, BLOCK_N, 0)]
+    constraints += [tkw.WaveConstraint(M, BLOCK_M)]
+    constraints += [tkw.WaveConstraint(N, BLOCK_N)]
+
+    @tkw.wave(constraints)
+    def test(
+        a: tkl.Memory[M, N, ADDRESS_SPACE, tkl.f16],
+        b: tkl.Memory[M, N, ADDRESS_SPACE, tkl.f16],
+        c: tkl.Memory[M, N, ADDRESS_SPACE, tkl.f16],
+    ):
+        lhs = tkw.read(a)
+        rhs = tkw.read(b)
+        tkw.debug_log_write(lhs)
+        tkw.debug_log_write(rhs)
+        res = lhs + rhs
+        tkw.debug_log_write(res)
+        tkw.write(res, c)
+
+    a = device_randn(shape, dtype=torch.float16)
+    b = device_randn(shape, dtype=torch.float16)
+    c = device_zeros(shape, dtype=torch.float16)
+    ref = a + b
+
+    options = WaveCompileOptions(
+        subs={
+            M: shape[0],
+            N: shape[1],
+            ADDRESS_SPACE: tkl.AddressSpace.GLOBAL_MEMORY.value,
+        },
+    )
+    options = set_default_run_config(options)
+
+    test = wave_compile(options, test)
+
+    debug_logs = {}
+    test(a, b, c, debug_logs=debug_logs)
+    assert_close(a, debug_logs["debug_log_output_0"])
+    assert_close(b, debug_logs["debug_log_output_1"])
+    assert_close(c, debug_logs["debug_log_output_2"])
```

Comment:
Could we somehow add a string tag to the `debug_log_write` op and use it for indexing here instead of a positional key? It will be annoying to have to update all keys back and forth when someone adds or removes a debug statement.

Reply:
I actually think assertions as you written may be an important use case too. One would want to compare with a reference, potentially after moving the data to a different device, and find diverging elements rather than eyeballing a potentially large tensor.